compiler/rustc_hir_analysis/src/check/method/suggest.rs

```diff
@@ -2324,6 +2324,61 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    /// issue #102320, for `unwrap_or` with closure as argument, suggest `unwrap_or_else`
+    /// FIXME: currently not working for suggesting `map_or_else`, see #102408
+    pub(crate) fn suggest_else_fn_with_closure(
+        &self,
+        err: &mut Diagnostic,
+        expr: &hir::Expr<'_>,
+        found: Ty<'tcx>,
+        can_satisfy: impl FnOnce(Ty<'tcx>) -> bool,
+    ) -> bool {
+        let Some((_def_id_or_name, output, _inputs)) = self.extract_callable_info(expr, found)
+        else { return false; };
+        if !can_satisfy(output) {
+            return false;
+        }
+
+        let parent = self.tcx.hir().get_parent_node(expr.hir_id);
+        if  let Some(Node::Expr(call_expr)) = self.tcx.hir().find(parent) &&
+            let hir::ExprKind::MethodCall(
+            hir::PathSegment { ident: method_name, .. },
+            self_expr,
+            args,
+            ..,
+        ) = call_expr.kind
+        {
+            let self_ty = self.typeck_results.borrow().expr_ty(self_expr);
```

Comment:
I think there is an `expr_ty_opt` version of this that doesn't ICE if the node is missing a type If so, then please use that, just in case.